src/bun.js/bindings/ZigGlobalObject.cpp

```diff
@@ -1779,6 +1784,75 @@ JSC_DEFINE_HOST_FUNCTION(jsReceiveMessageOnPort, (JSGlobalObject * lexicalGlobal
+    }
+
+    auto domain = arg0.toWTFString(globalObject);
+    if (domain.isNull())
+        return JSC::JSValue::encode(jsUndefined());
+    if (domain.is8Bit())
+        return JSC::JSValue::encode(arg0);
+
+    constexpr static int allowedNameToASCIIErrors = UIDNA_ERROR_EMPTY_LABEL | UIDNA_ERROR_LABEL_TOO_LONG | UIDNA_ERROR_DOMAIN_NAME_TOO_LONG | UIDNA_ERROR_LEADING_HYPHEN | UIDNA_ERROR_TRAILING_HYPHEN | UIDNA_ERROR_HYPHEN_3_4;
+    constexpr static size_t hostnameBufferLength = 2048;
+
+    auto encoder = &WTF::URLParser::internationalDomainNameTranscoder();
+    UChar hostnameBuffer[hostnameBufferLength];
+    UErrorCode error = U_ZERO_ERROR;
+    UIDNAInfo processingDetails = UIDNA_INFO_INITIALIZER;
+    int32_t numCharactersConverted = uidna_nameToASCII(encoder, StringView(domain).characters16(), domain.length(), hostnameBuffer, hostnameBufferLength, &processingDetails, &error);
+
+    if (U_SUCCESS(error) && !(processingDetails.errors & ~allowedNameToASCIIErrors) && numCharactersConverted) {
+        return JSC::JSValue::encode(JSC::jsString(vm, WTF::String(hostnameBuffer, numCharactersConverted)));
+    }
+    throwTypeError(globalObject, scope, "domainToASCII failed"_s);
+    return JSC::JSValue::encode(jsUndefined());
+}
+
+JSC_DEFINE_HOST_FUNCTION(functionDomainToUnicode, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    auto arg0 = callFrame->argument(0);
+    if (!arg0.isString()) {
+        throwTypeError(globalObject, scope, "the \"domain\" argument is missing or not a string"_s);
+        return JSC::JSValue::encode(jsUndefined());
+    }
+
+    auto domain = arg0.toWTFString(globalObject);
+    if (domain.isNull())
+        return JSC::JSValue::encode(jsUndefined());
+    if (!domain.is8Bit())
+        return JSC::JSValue::encode(arg0);
+
+    domain.convertTo16Bit();
+
+    constexpr static int allowedNameToUnicodeErrors = UIDNA_ERROR_EMPTY_LABEL | UIDNA_ERROR_LABEL_TOO_LONG | UIDNA_ERROR_DOMAIN_NAME_TOO_LONG | UIDNA_ERROR_LEADING_HYPHEN | UIDNA_ERROR_TRAILING_HYPHEN | UIDNA_ERROR_HYPHEN_3_4;
+    constexpr static int hostnameBufferLength = 2048;
+
+    auto encoder = &WTF::URLParser::internationalDomainNameTranscoder();
+    UChar hostnameBuffer[hostnameBufferLength];
+    UErrorCode error = U_ZERO_ERROR;
+    UIDNAInfo processingDetails = UIDNA_INFO_INITIALIZER;
+
+    int32_t numCharactersConverted = uidna_nameToUnicode(encoder, StringView(domain).characters16(), domain.length(), hostnameBuffer, hostnameBufferLength, &processingDetails, &error);
+
+    if (U_SUCCESS(error) && !(processingDetails.errors & ~allowedNameToUnicodeErrors) && numCharactersConverted) {
+        return JSC::JSValue::encode(JSC::jsString(vm, WTF::String(hostnameBuffer, numCharactersConverted)));
+    }
+    throwTypeError(globalObject, scope, "domainToUnicode failed"_s);
+    return JSC::JSValue::encode(jsUndefined());
```

Comment:
I like returning the `jsUndefined()` equiv when throwing in our zig code instead of `.zero` (which creates a nullptr backed JSValue which other JSValue methods get grumpy about). In looking at how WebKit handles this they are a bit all over the place but something I've seen that fits is at least returning the expected type. So in this case `jsEmptyString(vm)` which could be stored higher in the function like https://github.com/WebKit/WebKit/blob/e604d7c70127a9d0abd95421c65a2ca594df690e/Source/JavaScriptCore/runtime/JSCell.cpp#L252 then `JSC::JSValue::encode(emptyString)`.

Reply:
It is a bug that we return undefined instead of zero when there's an exception. This used to segfault. They added handling for it, but returning zero is good because it's a value which is unreachable from JavaScript. We cannot do any memory allocations while an exception is pending, they will fail. `jsEmptyString` might still work because the JSCell is not allocated on each call, but that is lucky/brittle

Reply:
> We cannot do any memory allocations while an exception is pending, they will fail.

Eeeh, that's a very good point.